### PR TITLE
Tide: Use a lister for ProwJobs

### PR DESCRIPTION
**Pull request by alvaroaleman**
This PR updates Tide to use a lister for Prowjobs. This is done to significantly reduce the cost of LIST operations against the kube api which we need in order to fix https://github.com/kubernetes/test-infra/issues/14233

### CODE REVIEW

**Comment by cblecker** [comment]
This seems sane to me, but I'd like another set of eyes on this
/cc @cjwagner

File: prow/cmd/tide/main.go [hunk 1]
```diff
@@ -182,8 +210,6 @@ func main() {
 		return
 	}
 	go func() {
-		sig := make(chan os.Signal, 1)
-		signal.Notify(sig, os.Interrupt, syscall.SIGTERM)
```
Comment (stevekuznetsov) [inline prow/cmd/tide/main.go]: This will conflict with #12927 :(

File: prow/tide/tide_test.go [hunk 2]
```diff
@@ -34,16 +34,22 @@ import (
...
 
 	prowapi "k8s.io/test-infra/prow/apis/prowjobs/v1"
-	"k8s.io/test-infra/prow/client/clientset/versioned/fake"
 	"k8s.io/test-infra/prow/config"
 	"k8s.io/test-infra/prow/git/localgit"
 	"k8s.io/test-infra/prow/github"
 	"k8s.io/test-infra/prow/tide/history"
 )
 
+func init() {
```
Comment (stevekuznetsov) [inline prow/tide/tide_test.go]: By putting the production variant of this into `main,go` and not an `init()` in `tide.go` aren't we _very technically_ making it incumbent on importers of Tide methods that use the client to do their own registration? Why don't we do the registration in an `init()` somewhere common?

File: prow/tide/tide_test.go [hunk 3]
```diff
@@ -1301,17 +1307,19 @@ func TestTakeAction(t *testing.T) {
...
+
 		var batchJobs []*prowapi.ProwJob
-		for _, action := range fakeProwJobClient.Actions() {
-			switch action := action.(type) {
-			case clienttesting.CreateActionImpl:
-				numCreated++
-				if prowJob, ok := action.Object.(*prowapi.ProwJob); ok && prowJob.Spec.Type == prowapi.BatchJob {
-					batchJobs = append(batchJobs, prowJob)
-				}
+		for _, pj := range prowJobs.Items {
```
Comment (stevekuznetsov) [inline prow/tide/tide_test.go]: Technically this isn't checking that they were created, right? Just that they exist in the end? So you could write a bad test where you accidentally add one from the start and not know?

File: prow/tide/tide_test.go (hunk 2 above)
Comment (alvaroaleman) [reply prow/tide/tide_test.go]: Good point, I could just add an init in the ProwJob api package, then we don't have to explicitly do that everywhere else?

File: prow/tide/tide_test.go (hunk 3 above)
Comment (alvaroaleman) [reply prow/tide/tide_test.go]: Yes, but I imagine doing that by accident is pretty hard, as the current table-driven design would then do that for all testcases

Comment (stevekuznetsov) [reply prow/tide/tide_test.go]: Does the controller-runtime fake even allow for checking the actions?

Comment (stevekuznetsov) [reply prow/tide/tide_test.go]: Does the controller-runtime fake even allow for checking the actions?

Comment (alvaroaleman) [reply prow/tide/tide_test.go]: Not to my knowledge, no.

Comment (stevekuznetsov) [reply prow/tide/tide_test.go]: Guess we'll just have to make do with this then

**Pull request merged**